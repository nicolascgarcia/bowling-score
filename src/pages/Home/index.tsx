import { ChangeEvent, FormEvent, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useGame } from '../../contexts/GameContext';
import { Container, NameInput, Button, UserName } from '../../components';
import { User } from '../../contexts/GameContext/gameContext.types';

export default function Home() {
	const navigate = useNavigate();

	const [userName, setUserName] = useState<string>('');
	const [users, setUsers] = useState<User[]>([]);

	const { addPlayers } = useGame();

	const removeUser = (index: number): void => {
		const currentUsers = [...users];
		currentUsers.splice(index, 1);
		setUsers(currentUsers);
	};

	const changeName = (event: ChangeEvent<HTMLInputElement>): void => {
		setUserName(event.target.value);
	};

	const handleSubmit = (event: FormEvent<HTMLFormElement>): void => {
		event.preventDefault();

		const user: User = {
			name: userName,
			photo: `https://avatars.dicebear.com/api/identicon/${userName}.svg`,
		};

		setUsers([...users, user]);
		setUserName('');
	};

	const startGame = () => {
		addPlayers(users);
		navigate('/score');
	};

	return (
		<Container
			bg='#0a0906'
			display='flex'
			justifyContent='center'
			alignItems='center'
			minHeight='100vh'
		>
			<Container
				width={['75vw', '50vw', '50vw', '30vw']}
				maxWidth='500px'
				bg='#121214'
				p={[4, 5, 5, 5]}
				borderRadius={10}
			>
				<form onSubmit={handleSubmit}>
					<Container display='flex'>
						<NameInput
							aria-label='name-input'
							type='text'
							placeholder="Add the player's name"
							value={userName}
							onChange={changeName}
							required
						/>
						<Container ml={1}>
							<Button type='submit' fontSize={28}>
								+
							</Button>
						</Container>
					</Container>
				</form>
				<Container my={3}>
					{users.map((user, index) => (
						<Container
							key={`${user.name}-${Math.random() * index}`}
							display='flex'
							alignItems='center'
						>
							<img src={user.photo} alt={user.name} width={50} height={50} />
							<Container ml={3} width={['60%', '65%', '65%', '65%']}>
								<UserName>{user.name}</UserName>
							</Container>
							<Container display='flex' flex={1} justifyContent='flex-end'>
								<Button
									type='button'
									onClick={() => removeUser(index)}
									fontSize={28}
								>
									-
								</Button>
							</Container>
						</Container>
					))}
				</Container>
				<Container>
					<Button type='button' width='100%' onClick={startGame}>
						start game
					</Button>
				</Container>
			</Container>
		</Container>
	);
}
