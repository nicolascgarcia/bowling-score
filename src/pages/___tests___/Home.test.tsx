/* eslint-disable import/no-extraneous-dependencies */
import { cleanup, fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Home from '../Home';

afterEach(cleanup);

describe('<Home/>', () => {
	it('should render Home', () => {
		const component = renderer.create(<Home />);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('should add user to list', async () => {
		const { getByRole } = render(<Home />);

		const input = getByRole('textbox', {
			name: 'name-input',
		}) as HTMLInputElement;
		const name = 'John';

		fireEvent.change(input, {
			target: { value: name },
		});

		expect(input.value).toBe(name);

		fireEvent.click(getByRole('button', { name: '+' }));

		expect(getByRole('heading', { name })).toBeVisible();
		expect(input.value).toBe('');
	});

	it('should remove user from list', async () => {
		const { getByRole, queryByRole } = render(<Home />);

		const input = getByRole('textbox', {
			name: 'name-input',
		}) as HTMLInputElement;
		const name = 'John';

		fireEvent.change(input, {
			target: { value: name },
		});

		expect(input.value).toBe(name);

		fireEvent.click(getByRole('button', { name: '+' }));

		expect(getByRole('heading', { name })).toBeVisible();
		expect(input.value).toBe('');

		fireEvent.click(getByRole('button', { name: '-' }));

		expect(queryByRole('heading', { name })).toBeNull();
	});
});
