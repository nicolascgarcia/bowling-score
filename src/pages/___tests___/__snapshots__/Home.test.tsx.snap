// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home/> should render Home 1`] = `
<div
  className="sc-bczRLJ ijiIxe"
  display="flex"
>
  <div
    className="sc-bczRLJ esRydr"
    width={
      Array [
        "75vw",
        "50vw",
        "50vw",
        "30vw",
      ]
    }
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="sc-bczRLJ btZVNS"
        display="flex"
      >
        <input
          aria-label="name-input"
          className="sc-gsnTZi hTKsgn"
          onChange={[Function]}
          placeholder="Add the player's name"
          required={true}
          type="text"
          value=""
        />
        <div
          className="sc-bczRLJ jIAPur"
        >
          <button
            className="sc-dkzDqf jAEYpy"
            fontSize={28}
            type="submit"
          >
            +
          </button>
        </div>
      </div>
    </form>
    <div
      className="sc-bczRLJ jFDCya"
    />
    <div
      className="sc-bczRLJ cBNlSK"
    >
      <button
        className="sc-dkzDqf kyfNmV"
        onClick={[Function]}
        type="button"
        width="100%"
      >
        start game
      </button>
    </div>
  </div>
</div>
`;
